seed-job/index_reader.py

```diff
@@ -0,0 +1,369 @@
+        process the seed-job template for same and oc apply changes
+        """
+        # get the oc process and oc apply command
+        command = self.oc_apply_seedjob_template_command()
+        # format the command with project params
+        command = command.format(
+            git_url=project.git_url,
+            git_path=project.git_path,
+            git_branch=project.git_branch,
+            target_file=project.target_file,
+            build_context=project.build_context,
+            desired_tag=project.desired_tag,
+            depends_on=project.depends_on,
+            notify_email=project.notify_email,
+            pipeline_name=project.pipeline_name,
+            app_id=project.app_id,
+            job_id=project.job_id,
+            pre_build_context=project.pre_build_context,
+            pre_build_script=project.pre_build_script,
+            registry_url=self.registry_url,
+            from_address=self.from_address,
+            smtp_server=self.smtp_server
+        )
+        # process and apply buildconfig
+        output = run_cmd(command, shell=True)
+        print (output)
+
+        # if a buildConfig has config update, oc apply returns
+        # "buildconfig.build.openshift.io "$PIPELINE_NAME" configured"
+        # possible values are ["unchanged", "created", "configured"]
+        # we are looking for "configured" string for updated pipelines
+        if "configured" in output:
+            print ("{} is updated, starting build..".format(
+                project.pipeline_name))
+            self.start_build(project.pipeline_name)
+
+    def start_build(self, pipeline_name):
+        """
+        Given a pipeline name, start the build for same
+        """
+        command = "oc start-build {} -n {}".format(
+            pipeline_name, self.namespace)
+        print (run_cmd(command))
+
+    def delete_buildconfigs(self, bcs):
+        """
+        Deletes the given list of bcs
+        """
+        command = "oc delete -n {} bc {}"
+        for bc in bcs:
+            print ("Deleting buildConfig {}".format(bc))
+            run_cmd(command.format(self.namespace, bc))
+
+
+class Index(object):
+    """
+    The orchestrator class to process operations
+    in container index.
+    """
+
+    def __init__(self, index, registry_url, namespace,
+                 from_address, smtp_server):
+        # create index reader object
+        self.index_reader = IndexReader(index, namespace)
+        # create dc_manager object
+        self.dc_manager = DeploymentConfigManager(
+            registry_url, namespace, from_address, smtp_server)
+        self.infra_projects = ["seed-job"]
+
+    def find_stale_jobs(self, oc_projects, index_projects):
+        """
+        Given oc projects and index projects, figure out stale
+        projects and return
+        """
+        # diff existing oc projects from index projects
+
+        return list(set(oc_projects) - set(index_projects))
```

Comment:
I am guessing we're using `set()` here to make sure that the values are unique, right? But doesn't OpenShift take care of it already? Would it let two buildconfigs with exact same name to exist in the same namespace? I think no.

Reply:
We are it using to do `-` operation, not for uniqueness.